### DISCUSSION TITLE
recursive collections?  category -> parent (category)

### DISCUSSION THREAD

**Discussion by movingelectrons** [pocketbase/pocketbase#5067] [Q&A]
Hello,

Having fun playing with pocketbase as the backend for my project, going well so far ablit a tad slow but I am brand new to go.

I have a collection named 'categories' (cats for short)

is can optionally have a parent, which is also a cat, calling that 'parent'.   entering data into this is just fine,

I can have categories food, nonfood, inside food is a cat veggies (parent is food) and fruits, parent is also food.  inside nonfood i have category cars, inside cars is category volvo, chevy, etc, you get the idea.

How can I get the data out of this?  I have the first level done thats easy, just filter where parent is blank.  But everything else I'm having a hard time figuring out how to do it without a million calls to the database, I would like to get the first 3 levels

**Comment by ganigeorgiev** [comment]
If the queries are executed as part of a hook or custom route (aka. on the PocketBase server) then it may be easier to just perform the individual recursive db calls ([Many Small Queries Are Efficient In SQLite](https://sqlite.org/np1queryprob.html)).

If you know that the collection will have at least 3 levels and you want to get them with one Web API call, then you could use expand like:
```js
const cats = await pb.collection("categories").getFullList({
    // see https://pocketbase.io/docs/working-with-relations/#back-relations
    expand: "categories_via_parent.categories_via_parent.categories_via_parent
})
``` 

Another option could be to create a View collection and a query with [recursive CTE](https://www.sqlite.org/lang_with.html) (_or patterns like [nested set](https://en.wikipedia.org/wiki/Nested_set_model) but I don't recommend it as they usually require manually managing special columns_).
There is no information about your collection or what actually you want the "tree" to look like, nor how do you plan to use it, but you can find an example query in https://github.com/pocketbase/pocketbase/discussions/2769.

_(edit: replaced the expand example to use back relations)_

**Comment by movingelectrons** [comment]
trying to generate a list of the main cats and under each main cat list the cats that belong to them

following the example i gave would get a list like

1. food
  a. veggies
  b. fruits
2. nonfood
  a. cars
    1. volvo
    2. chevy

**Reply by ganigeorgiev** [reply]
If you want the root/top-level categories with their children nested then in that case using `expand` should work:

```js
const cats = await pb.collection("categories").getFullList({
    filter: "parent = ''", // fetch only roots
    // expand root children
    // (see https://pocketbase.io/docs/working-with-relations/#back-relations)
    expand: "categories_via_parent.categories_via_parent.categories_via_parent"
})
```

**Comment by movingelectrons** [comment]
i cant get it to work.  and I have to use .getList(1,50{  for some reason .getFullList dosen't work..

**Reply by ganigeorgiev** [reply]
I don't understand what do you mean by this.

`getFullList` calls `getList` until all items are fetched.

If there is an error, please post it otherwise I don't think I can help further.

If the error is on the server-side you can also try to start the executable with `--dev` and/or check the logs in the Admin UI > Logs.

**Reply by movingelectrons** [reply]
i got it figured out, just made two calls, and made two separate tables, one for categories and one for subcategories, i just get the full list of both of them and put them together with javascript.
